### PR TITLE
Fixes #2454 Added Readability score and average sentence length text property

### LINKED ISSUE #2454: [FEAT] Additional NLP properties (2)
Add additional properties for the NLP sub package.

We have a lot of checks in the NLP sub package that rely on high quality text properties.
Currently we have some which can be seen at text_properties.py.

We need to add more, can be based on simple internal python methods like text length or complex that uses external models such as formality.

When suggesting a new property - include what is the run time for the property and what are the dependencies it has

### PR DESCRIPTION
**Pull request by hjain5164**
Fixes #2454 

#### What does this implement/fix? Explain your changes.
I have added two new text properties:
~~Automated Readability Index:  A score that represents the grade level required to understand the text.~~
- Average Sentence Length: Average number of words per sentence in the text.
- Readability Score: Returns a score as per the Flesch reading-ease test, higher scores indicate material that is easier to read whereas lower numbers mark texts that are more difficult to read.

After discussion, removed ARI based readability score and added the Flesch Reading-Ease score since that would work better for the short and long forms of texts.

Time is taken for the two metrics on the tweet dataset consisting of 4653 samples (in seconds): 
- Average Sentence Length: 01.225178 seconds
~~Automated Readability Index: 00.356193 seconds~~
- Readability Score: 01.428470 seconds

#### Any other comments?
Added the test to calculate these two new properties on the tweet test dataset and asserted with the first 10 results.

### CODE REVIEW

File: deepchecks/nlp/utils/text_properties.py [hunk 1]
```diff
@@ -335,6 +338,57 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+        warnings.warn('nltk punkt not found, automated readability index cannot be calculated.'
+                      ' Please check your internet connection.', UserWarning)
+        return [np.nan] * len(raw_text)
+    result = []
+    for text in raw_text:
+        if not pd.isna(text):
+            sentence_count = len(sent_tokenize(text))
+            word_count = len(re.findall(r'\w+', text))
+            char_count = len(re.sub(r'\s', '', text))
+            try:
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:359]: try catch is slow, can replace it with:
if word_count ==0 or sentence_count==0:
result.append(1) # smallest ari index
else:
...

File: deepchecks/nlp/utils/text_properties.py [hunk 2]
```diff
@@ -335,6 +338,57 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
     return result
 
 
+def automated_readability_index(raw_text: Sequence[str]) -> List[str]:
```
[the remark this answers is not in the document]
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: deepchecks/nlp/utils/text_properties.py [hunk 3]
```diff
@@ -335,6 +338,57 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+    if not nltk_download('punkt', quiet=True):
+        warnings.warn('nltk punkt not found, average sentence length cannot be calculated.'
+                      ' Please check your internet connection.', UserWarning)
+        return [np.nan] * len(raw_text)
+    result = []
+    for text in raw_text:
+        if not pd.isna(text):
+            sentences = [remove_punctuation(sent) for sent in sent_tokenize(text)]
+            total_words = sum([len(word_tokenize(sentence)) for sentence in sentences])
+            try:
```
[the remark this answers is not in the document]
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: docs/source/nlp/usage_guides/nlp_properties.rst [hunk 4]
```diff
@@ -59,6 +59,8 @@ Fluency*                        Fluency of the text. Uses the prithivida/parrot_
 Formality*                      Formality of the text. Uses the s-nlp/roberta-base-formality-ranker model
 Lexical Density                 Percentage of unique words in the text, rounded up to 2 decimal digits
 Unique Noun Count*              Number of unique noun words in the text
+Automated Readability Index     A score that represents the grade level required to understand the text.
```
[the remark this answers is not in the document]
Comment (hjain5164) [reply docs/source/nlp/usage_guides/nlp_properties.rst]: Done

File: deepchecks/nlp/utils/text_properties.py [hunk 5]
```diff
@@ -335,6 +338,57 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+    for text in raw_text:
+        if not pd.isna(text):
+            sentence_count = len(sent_tokenize(text))
+            word_count = len(re.findall(r'\w+', text))
+            char_count = len(re.sub(r'\s', '', text))
+            try:
+                a = float(char_count) / float(word_count)
+                b = float(word_count) / float(sentence_count)
+                ari = ((4.71 * round(a, 2)) + (0.5 * round(b, 2)) - 21.43)
+                result.append(round(ari, 0))
```
[the remark this answers is not in the document]
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: deepchecks/nlp/utils/text_properties.py (hunk 1 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: deepchecks/nlp/utils/text_properties.py [hunk 6]
```diff
@@ -335,6 +338,63 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+        words = word_tokenize(text)
+        word_count = len(words)
+        syllable_count = sum([len(cmudict_corpus[word.lower()]) for word in words if word.lower() in cmudict_corpus])
+        if word_count != 0 and sentence_count != 0 and syllable_count != 0:
+            avg_syllables_per_word = syllable_count / word_count
+            avg_words_per_sentence = word_count / sentence_count
+            flesch_reading_ease = 206.835 - (1.015 * avg_words_per_sentence) - (84.6 * avg_syllables_per_word)
+            result.append(round(flesch_reading_ease, 1))
+        else:
+            result.append(0.0)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:373]: I think the appropriate score here is to set None

File: deepchecks/nlp/utils/text_properties.py [hunk 7]
```diff
@@ -335,6 +338,63 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+    """Return a list of floats of Flesch Reading-Ease score per text sample.
+
+    In the Flesch reading-ease test, higher scores indicate material that is easier to read
+    whereas lower numbers mark texts that are more difficult to read. For more information:
+    https://en.wikipedia.org/wiki/Flesch%E2%80%93Kincaid_readability_tests#Flesch_reading_ease
+    """
+    if not nltk_download('punkt', quiet=True):
+        warnings.warn('nltk punkt not found, readability score cannot be calculated.'
+                      ' Please check your internet connection.', UserWarning)
+        return [0] * len(raw_text)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:351]: also here set None instead of 0

File: deepchecks/nlp/utils/text_properties.py [hunk 8]
```diff
@@ -335,6 +338,63 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+    result = []
+    for text in raw_text:
+        if not pd.isna(text):
+            sentences = [remove_punctuation(sent) for sent in sent_tokenize(text)]
+            total_words = sum([len(word_tokenize(sentence)) for sentence in sentences])
+            if len(sentences) != 0:
+                asl = round(total_words / len(sentences))
+                result.append(round(asl, 0))
+            else:
+                result.append(0.0)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:392]: same here set None

File: deepchecks/nlp/utils/text_properties.py [hunk 9]
```diff
@@ -335,6 +338,63 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+        sentence_count = len(sent_tokenize(text))
+        text = remove_punctuation(text)
+        words = word_tokenize(text)
+        word_count = len(words)
+        syllable_count = sum([len(cmudict_corpus[word.lower()]) for word in words if word.lower() in cmudict_corpus])
+        if word_count != 0 and sentence_count != 0 and syllable_count != 0:
+            avg_syllables_per_word = syllable_count / word_count
+            avg_words_per_sentence = word_count / sentence_count
+            flesch_reading_ease = 206.835 - (1.015 * avg_words_per_sentence) - (84.6 * avg_syllables_per_word)
+            result.append(round(flesch_reading_ease, 1))
```
[the remark this answers is not in the document]
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py:371]: I think we should, else the float operations may make the result go till many decimal points.

[the remark this answers is not in the document]
Comment (Nadav-Barak) [reply deepchecks/nlp/utils/text_properties.py]: sure, round to , 3 than

File: deepchecks/nlp/utils/text_properties.py [hunk 10]
```diff
@@ -335,6 +338,63 @@ def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
...
+            sentence_count = len(sent_tokenize(text))
+            text = remove_punctuation(text)
+            words = word_tokenize(text)
+            word_count = len(words)
+            syllable_count = sum([len(cmudict_dict[word.lower()]) for word in words if word.lower() in cmudict_dict])
+            if word_count != 0 and sentence_count != 0 and syllable_count != 0:
+                avg_syllables_per_word = syllable_count / word_count
+                avg_words_per_sentence = word_count / sentence_count
+                flesch_reading_ease = 206.835 - (1.015 * avg_words_per_sentence) - (84.6 * avg_syllables_per_word)
+                result.append(round(flesch_reading_ease, 1))
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:369]:
```suggestion
                result.append(round(flesch_reading_ease, 3))
```

**Review by Nadav-Barak — APPROVED** [review]
Great work!

File: deepchecks/nlp/utils/text_properties.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done. Pushing few changes for tests since now we are now rounding it off to 3 decimal places.

**Review by noamzbr — APPROVED** [review]

**Pull request merged**